File: problems/arrays/min_rewards/python/min_rewards.py
def minRewards(scores: list[int]) -> int:
    n = len(scores)
    rewards = [1] * n

    def traverse(startIdx: int, endIdx: int, step: int) -> None:
        for i in range(startIdx, endIdx, step):
            if scores[i] > scores[i - step]:
                rewards[i] = (
                    rewards[i - step] + 1
                    if step == 1
                    else max(rewards[i], rewards[i - step] + 1)
                )

    traverse(1, n, 1)
    traverse(n - 2, -1, -1)

    return sum(rewards)

File: problems/arrays/min_rewards/js/min_rewards.js
export const minRewards = (scores) => {
  const n = scores.length;
  const rewards = new Array(n).fill(1);

  const traverse = (startIdx, endIdx, step) => {
    for (let i = startIdx; i !== endIdx; i += step) {
      if (scores[i] > scores[i - step]) {
        rewards[i] =
          step == 1
            ? rewards[i - step] + 1
            : Math.max(rewards[i], rewards[i - step] + 1);
      }
    }
  };

  traverse(1, n, 1);
  traverse(n - 2, -1, -1);

  return rewards.reduce((sum, reward) => sum + reward, 0);
};
